rs/tests/nns/rent_subnet_test.rs

```diff
@@ -0,0 +1,560 @@
+                        node.node_id, rented_subnet_id, err,
+                    );
+                })
+        });
+
+    let a_rented_subnet_node = rented_subnet
+        .nodes()
+        .next()
+        .expect("Could not find any node in newly created subnet.");
+    let mut agent = assert_create_agent(a_rented_subnet_node.get_public_url().as_str()).await;
+    agent.set_identity(
+        BasicIdentity::from_pem(std::io::Cursor::new(SUBNET_USER_KEYPAIR.to_pem())).unwrap(),
+    );
+
+    // Verify 2 & 3: Can create a canister, and install code into it.
+    let new_canister =
+        UniversalCanister::new_with_retries(
+            &agent,
+            a_rented_subnet_node.effective_canister_id(),
+            // DO NOT MERGE 1_000_000_000_000_000, // cycles. 1e15, i.e. 1 trillion
+            log,
+        )
+        .await;
+    let new_canister_principal_id = CanisterId::unchecked_from_principal(
+        PrincipalId::from(new_canister.canister_id())
+    );
+
+    // Verify 2.1: The created canister is actually IN the rented subnet, due to
+    // it being created by the rented subnet's user.
+    assert_canister_belongs_to_subnet(
+        topology_snapshot,
+        new_canister_principal_id,
+        rented_subnet_id,
+    ).await;
+
+    // This will be used later to verify 6: the canister does not charged cycles.
+    let original_cycles_balance = get_cycles_balance(
+        new_canister_principal_id,
+        &rented_subnet,
+    ).await;
+
+    // Verify 4: Can call the canister.
+    const UPDATE_MSG_1: &[u8] =
+        b"This beautiful prose should be persisted for future generations";
+    new_canister.store_to_stable(0, UPDATE_MSG_1).await;
+
+    // Verify 5: The previous call had the intended effect.
+    assert_eq!(
+        new_canister
+            .try_read_stable(0, UPDATE_MSG_1.len() as u32)
+            .await,
+        UPDATE_MSG_1.to_vec(),
+    );
+
+    // Verify 6: The canister was not charged for the previous two operations.
+    let later_cycles_balance = get_cycles_balance(
+        new_canister_principal_id,
+        &rented_subnet,
+    ).await;
+    assert_eq!(
+        later_cycles_balance - original_cycles_balance,
+        0,
+        "{original_cycles_balance}",
+    );
+
+    // Verify 7: Other principals are NOT allowed to create canisters in the
+    // rented subnet.
+
+    // Switch to another user
+    // != is used here, bcause the operands do not implement Debug.
+    assert!(*SUBNET_USER_KEYPAIR != *TEST_USER2_KEYPAIR);
+    agent.set_identity(
+        BasicIdentity::from_pem(std::io::Cursor::new(TEST_USER2_KEYPAIR.to_pem())).unwrap(),
+    );
+
+    // As another user, try to create a canister in the rented subnet. This is supposed to get blocked.
+    let create_canister_result = UniversalCanister::try_new(
+        &agent,
+        a_rented_subnet_node.effective_canister_id(),
+    )
+    .await;
+    assert!(create_canister_result.is_err());
```

Comment:
This last assert is failing for me. So, canister creation succeeds, I guess. I guess my question here is, does this canister creation take place via CMC? Doesn't it have to?